dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java

```diff
@@ -2416,7 +2416,13 @@ public int saveTaskRelation(User operator, long projectCode, long processDefinit
         Map<Long, TaskDefinitionLog> taskDefinitionLogMap = null;
         if (CollectionUtils.isNotEmpty(taskDefinitionLogs)) {
             taskDefinitionLogMap = taskDefinitionLogs.stream()
-                    .collect(Collectors.toMap(TaskDefinition::getCode, taskDefinitionLog -> taskDefinitionLog));
+                    .collect(
+                        Collectors.toMap(
+                            TaskDefinition::getCode,
+                            taskDefinitionLog -> taskDefinitionLog,
+                            (entityOld , entityNew) -> entityOld
```

Comment:
I am actually not sure the logic here, but I guess you want to retain the latest `taskDefinitionLog` in the map?
If so, you bettered compared by id or operateTime to determine which should retain.

Reply:
![](https://vip2.loli.io/2022/02/08/vsGeKqj5m48gBS7.png)

In the production environment where I upgraded from 2.0.2-release to 2.0.3-release, when entering the workflow definition modification content from the workflow instance, the front end would send two identical data, which caused this error. However, I am currently unable to reproduce this problem in standalone mode. It makes me confused.

Reply:
@JinyLeeChina plz check this

Reply:
![](https://vip2.loli.io/2022/02/11/R7bASNIcOM3VepJ.png)

After several days of careful investigation, I probably found the cause of this situation.

Steps to reproduce:
1.Create a workflow with dependent nodes in 1.3.9, and the dependent nodes contain more than 1 dependency.
2.Create a new process definition by using copy workflow and then modify copyed workflow content randomly.
3.Upgrade version from 1.3.9 to 2.0.3.
4.Update process instance/definition.

The reason is that after using the copy workflow, the dependency will be copied, resulting in multiple identical pre tasks. At this time, there will be multiple workflow definitions and workflow relationships in the request sent by saving the workflow definition.

Reply:
> @JinyLeeChina plz check this. i am not sure its as same as the issue #8043.
> 
> the logic maybe like this according to @JinyLeeChina .
> 
> any part of the workflow changed would lead to the workflow's version change. so when you changed one taskDefinition, the taskDefinition has two version now, but the workflow doesn`t. When ui query this workflow it returns two taskDefinition with same code but different version and this cause the duplicate error

Yes, I also think this is the same problem as #8043. It has been fundamentally solved in version 203. There are design changes from 202 to 203. From 2.0.0 to 2.0.2, the version is designed with task as the core, and the version is designed with workflow as the core after 2.0.3. Therefore, the PR submitted in #8043 is for this design change. At the same time, the problem of repeated key is also fundamentally solved, not just the bug